### ISSUE TITLE
Can you provide the codes for training VITs on Imagenet?

### ISSUE THREAD

**Issue by 123123qqqqdqd** [muellerm-95/SAM-BN#2]
Thank you very much!

**Comment by mueller-mp** [comment]
Hi, I unfortunately cannot share the full code, but provide an example that should be easy to modify and contain the relevant logic, i.e. the optimizer and how the normalization layers are selected. Here is the LION optimizer, which worked best in our experiments when combined with SAM-ON.

```
from torch import Tensor
import torch
import math
import torchvision
import torchvision.transforms as transforms
import torch.nn.functional as F

# create separate param group for normalization layers
def add_sam(model, only_norm=False):
    sam = []
    no_sam = []
    for name, param in model.named_parameters():
        if not param.requires_grad:
            continue  # frozen weights
        if (('bn' not in name) and ('norm' not in name)) and only_norm:
            no_sam.append(param)
        else:
            sam.append(param)
    return [group for group in
            [
                {'params': no_sam, 'sam': False},
                {'params': sam, 'sam': True}
            ]
            if len(group['params']) > 0]


# adapted from https://github.com/google/automl/blob/master/lion/lion_pytorch.py
class Lion_with_SAM(torch.optim.Optimizer):
    def __init__(
            self, params, lr=0.001, betas=(0.9, 0.99), weight_decay=1.,rho=10.,clip_norm=1.
            ):
        defaults = dict(
            lr=lr,
            betas=betas,
            weight_decay=weight_decay,
            rho=rho,
        )
        super(Lion_with_SAM, self).__init__(params, defaults)
        self.clip_norm=clip_norm
        # "exp_avg" has to exist for lion_step
        for group in self.param_groups:
            for p in group['params']:
                if p.requires_grad:
                    if 'exp_avg' not in self.state[p]:
                        self.state[p]["exp_avg"] = torch.zeros_like(p).detach()


    @torch.no_grad()
    def perturb(self):
        gradient_norm = torch.norm(torch.stack([p.grad.norm(p=2) for group in self.param_groups 
                                                for p in group["params"] 
                                                if (p.grad is not None and group['sam'])]),p=2) + 1e-10
        for group in self.param_groups:
            for p in group["params"]:
                if p.grad is None or not group['sam']:
                    continue
                epsilon = group['rho']*p.grad / (gradient_norm) 
                p.add_(epsilon)
                self.state[p]["epsilon"] = epsilon.detach().clone()

    @torch.no_grad()
    def unperturb(self):
        for group in self.param_groups:
            for p in group["params"]:
                if p.grad is None or not group['sam']:
                    continue
                p.sub_(self.state[p]["epsilon"])

    # the actual lion update
    @torch.no_grad()
    def lion_step(self):
        for group in self.param_groups:
            for p in group['params']:
                if p.grad is None:
                    continue

                # Perform stepweight decay
                p.data.mul_(1 - group['lr'] * group['weight_decay'])

                grad = p.grad
                state = self.state[p]
                # State initialization
                if len(state) == 0:
                    # Exponential moving average of gradient values
                    state['exp_avg'] = torch.zeros_like(p)

                exp_avg = state['exp_avg']
                beta1, beta2 = group['betas']

                # Weight update
                update = exp_avg * beta1 + grad * (1 - beta1)
                p.add_(torch.sign(update), alpha=-group['lr'])
                # Decay the momentum running average coefficient
                exp_avg.mul_(beta2).add_(grad, alpha=1 - beta2)

    def step(self, model, x, y, loss_fct):
        # first forwad-backward to compute perturbation
        preds = model(x)
        loss = loss_fct(preds, y)
        loss.mean().backward()
        self.perturb()
        self.zero_grad()

        # second forward-backward to compute gradient for actual update
        loss_2 = loss_fct(model(x), y)
        loss_2.mean().backward()
        self.unperturb() # move back to original weights

        # if desired, clip gradient norm here
        if self.clip_norm is not None:
            torch.nn.utils.clip_grad_norm_(model.parameters(), self.clip_norm)

        # perform actual update according to lion optimizer
        self.lion_step()
        self.zero_grad()
```
And the ImageNet training works as usual, you would have to add lr decay, logging, validation etc. yourself:
```
# hyperparameters
lr=0.001
weight_decay=1.
rho=10.
clip_norm=1.
only_norm=True
num_epochs=300
batch_size = 1024

# data
tfs = transforms.Compose([transforms.RandomResizedCrop(224), transforms.RandomHorizontalFlip(), transforms.ToTensor(), transforms.Normalize(mean=[0.485, 0.456, 0.406],std=[0.229, 0.224, 0.225])])
imagenet_data = torchvision.datasets.ImageNet('/path/to/imagenet/',split='train',transform=tfs)
data_loader = torch.utils.data.DataLoader(imagenet_data,batch_size=batch_size,shuffle=True,num_workers=8)

# model and optimizer
model = ... # add model here, make sure that ALL normalization layers contain the string 'norm', you'll potentially have to adjust the model definition for this
model = model.cuda()
params = add_sam(model,only_norm=only_norm) 
optimizer = Lion_with_SAM(params=params,lr=lr, weight_decay=weight_decay, rho=rho, clip_norm=clip_norm)
loss_fct = F.cross_entropy

for epoch in range(num_epochs):
    # add lr decay, logging, validation ...
    for i,(x,y) in enumerate(data_loader):
        x, y = x.cuda(),y.cuda()
        optimizer.step(model, x, y, loss_fct)

```
Hope this helps!

**Comment by 123123qqqqdqd** [comment]
Thank you for your reply! I notice that you set rho=10, which seems extremely large. Is this referring to adaptive version or original version? Is this rho=10 used for SAM-ON VIT training? May I ask what is the corresponding rho you used for SAM and ASAM when training VITs? I am sorry that I am not familiar with VIT training using SAMs. Thank you very much!

**Comment by mueller-mp** [comment]
$\rho=10$ is for the non-adaptive SAM-ON version. SAM-ON typically requires larger perturbation radii than SAM. You can find the detailed hyperparameters in Table 8 of the paper:
https://arxiv.org/pdf/2306.04226.pdf

For Lion, the reported values of $\rho$ were: 
$\rho=1$ for SAM
$\rho=10$ for SAM-ON

**Comment by mueller-mp** [comment]
Since there seem to be no more questions, I will close this issue. Feel free to re-open it, in case anything is unclear. 
Regarding code for training SAM on ImageNet, [this](https://github.com/MarlonBecker/MSAM) codebase was recently published and might be helpful.
